src/gcp.rs

```diff
@@ -199,24 +199,29 @@ impl GcpAuthenticator {
         match self {
             Self::Credentials(inner) => {
                 let expires_in = inner.token.read().unwrap().expires_in() as u64;
-                let mut deadline =
-                    Duration::from_secs(expires_in - METADATA_TOKEN_EXPIRY_MARGIN_SECS);
+                let mut deadline = Duration::from_secs(
+                    expires_in
+                        .saturating_sub(METADATA_TOKEN_EXPIRY_MARGIN_SECS)
+                        .max(1),
+                );
                 loop {
+                    debug!(
+                        "Sleeping for {}s before refreshing GCP authentication token",
+                        deadline.as_secs()
+                    );
                     tokio::time::sleep(deadline).await;
-                    debug!("Renewing GCP authentication token.");
                     match inner.regenerate_token().await {
                         Ok(()) => {
                             sender.send_replace(());
+                            debug!("GCP authentication token renewed.");
                             let expires_in = inner.token.read().unwrap().expires_in() as u64;
                             // Rather than an expected fresh token, the Metadata Server may return
                             // the same (cached) token during the last 300 seconds of its lifetime.
                             // This scenario is handled by retrying the token refresh after the
                             // METADATA_TOKEN_ERROR_RETRY_SECS period when a fresh token is expected
-                            let new_deadline = if expires_in <= METADATA_TOKEN_EXPIRY_MARGIN_SECS {
-                                METADATA_TOKEN_ERROR_RETRY_SECS
-                            } else {
-                                expires_in - METADATA_TOKEN_EXPIRY_MARGIN_SECS
-                            };
+                            let new_deadline = expires_in
+                                .saturating_sub(METADATA_TOKEN_ERROR_RETRY_SECS)
+                                .max(1);
```

Comment:
This is wrong. Be careful not to mix up `ERROR_RETRY_SECS` with `EXPIRY_MARGIN_SECS`
```suggestion
                            let new_deadline = expires_in
                                .saturating_sub(METADATA_TOKEN_ERROR_MARGIN_SECS)
                                .max(METADATA_TOKEN_ERROR_RETRY_SECS);
```

Reply:
Nice catch